components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/graphQL/GraphQLQueryAnalysisHandler.java

```diff
@@ -0,0 +1,353 @@
+import org.apache.axiom.om.OMFactory;
+import org.apache.axiom.om.OMNamespace;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpStatus;
+import org.apache.synapse.Mediator;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.rest.AbstractHandler;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.wso2.carbon.apimgt.gateway.handlers.Utils;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityUtils;
+import org.wso2.carbon.apimgt.gateway.handlers.security.usermgt.APIKeyMgtRemoteUserClient;
+import org.wso2.carbon.apimgt.gateway.handlers.security.usermgt.APIKeyMgtRemoteUserClientPool;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+
+public class GraphQLQueryAnalysisHandler extends AbstractHandler {
+
+    private static final Log log = LogFactory.getLog(GraphQLQueryAnalysisHandler.class);
+    private GraphQLSchema schema = null;
+
+    public boolean handleRequest(MessageContext messageContext) {
+        schema = (GraphQLSchema) messageContext.getProperty(APIConstants.GRAPHQL_SCHEMA);
+        String payload = messageContext.getProperty(APIConstants.GRAPHQL_PAYLOAD).toString();
+        if (!analyseQuery(messageContext, payload)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Query was blocked by the static query analyser");
+            }
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * This method returns the user roles
+     *
+     * @param username username of the user
+     * @return list of user roles
+     */
+    private String[] getUserRoles(String username) throws APISecurityException {
+        String[] userRoles;
+        APIKeyMgtRemoteUserClient client;
+        try {
+            APIKeyMgtRemoteUserClientPool clientPool = APIKeyMgtRemoteUserClientPool.getInstance();
+            client = clientPool.get();
+            userRoles = client.getUserRoles(username);
+        } catch (Exception e) {
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR, e.getMessage(), e);
+        }
+        return userRoles;
+    }
+
+    /**
+     * This method returns the maximum query depth value
+     *
+     * @param userRoles        list of user roles
+     * @param policyDefinition json object which contains the policy
+     * @return maximum query depth value if exists, or -1 to denote no depth limitation
+     */
+    private int getMaxQueryDepth(String[] userRoles, JSONObject policyDefinition) {
+        Object depthObject = policyDefinition.get(APIConstants.QUERY_ANALYSIS_DEPTH);
```

Comment:
shouldn't we check policyDefinition.has(APIConstants.QUERY_ANALYSIS_DEPTH) before trying to get it ?
applies to L104 also

Reply:
Not applicable anymore since the master branch has a different codebase.